router/internal/circuit/breaker.go

```diff
@@ -0,0 +1,65 @@
+package circuit
+
+import (
+	"context"
+	"fmt"
+	"github.com/wundergraph/cosmo/router/internal/traceclient"
+	"go.uber.org/zap"
+	"net/http"
+)
+
+type Breaker struct {
+	roundTripper   http.RoundTripper
+	loggerFunc     func(req *http.Request) *zap.Logger
+	circuitBreaker *Manager
+}
+
+func NewCircuitTripper(roundTripper http.RoundTripper, breaker *Manager, logger func(req *http.Request) *zap.Logger) *Breaker {
+	return &Breaker{
+		circuitBreaker: breaker,
+		loggerFunc:     logger,
+		roundTripper:   roundTripper,
+	}
+}
+
+func (rt *Breaker) RoundTrip(req *http.Request) (resp *http.Response, err error) {
+	ctx := req.Context()
+
+	var subgraph string
+	subgraphCtxVal := ctx.Value(traceclient.CurrentSubgraphContextKey{})
+	if subgraphCtxVal != nil {
+		subgraph = subgraphCtxVal.(string)
+	}
+
+	var prefix string
+	prefixCtxVal := ctx.Value(traceclient.CurrentFeatureFlagContextKey{})
+	if prefixCtxVal != nil {
+		prefix = prefixCtxVal.(string)
+	}
+
+	cbKey := fmt.Sprintf("%s::%s", prefix, subgraph)
+
+	// If there is no circuit defined for this subgraph
+	circuit := rt.circuitBreaker.GetCircuitBreaker(cbKey)
+	if circuit == nil {
+		return rt.roundTripper.RoundTrip(req)
+	}
+
+	preRunStatus := circuit.IsOpen()
+
+	err = circuit.Run(context.Background(), func(_ context.Context) error {
```

Comment:
To check if we should not use `context.Background()` here

Reply:
Why should the cb not run within the request context? If you pass req ctx, you also might to check for `ctx.Err()` inside the callback right?